src/Xrefcheck/Verify.hs

```diff
@@ -327,5 +370,62 @@ checkExternalResource VerifyConfig{..} link
         HttpExceptionRequest _ exc -> case exc of
           StatusCodeException resp _
             | isAllowedErrorCode (statusCode $ responseStatus resp) -> Right ()
-            | otherwise -> Left $ ExternalResourceUnavailable (responseStatus resp)
+            | otherwise -> Left $ ExternalHttpResourceUnavailable (responseStatus resp)
           other -> Left . ExternalResourceSomeError $ show other
+
+    checkFtp :: URI -> Bool -> ExceptT VerifyError IO ()
+    checkFtp uri secure = do
+      -- get authority which stores host and port
+      authority <- case uriAuthority uri of
+        Right a -> pure a
+        Left _ -> throwError $
+          ExternalResourceInvalidUrl (Just "FTP path must be absolute")
+      let host = toString $ authHost authority
+      port :: Int <- case toIntegralSized . fromMaybe 21 $ authPort authority of
+        Just p -> pure p
+        Nothing -> throwError $
+          ExternalResourceInvalidUrl (Just "Bad port")
+      -- build path from pieces
+      path <- case uriPath uri of
+        Nothing -> pure ""
+        Just (_, pieces) -> pure
+          . mconcat
+          . intersperse "/"
+          . map toString
+          . toList
+          $ pieces
+      makeFtpRequest host port path secure `catch` \e ->
+        throwError $ ExternalFtpException e
+
+    makeFtpRequest
+      :: String
+      -> Int
+      -> FilePath
+      -> Bool
+      -> ExceptT VerifyError IO ()
+    makeFtpRequest host port path secure = handler host port $
+      \handle response -> do
+        -- check connection status
+        when (frStatus response /= Success) $
+          throwError $ ExternalFtpResourceUnavailable response
+        -- anonymous login
+        loginResp <- login handle "anonymous" ""
+        -- check login status
+        when (frStatus loginResp /= Success) $
+          if vSkipProtected
+          then pure ()
+          else throwError $ ExternalFtpException $ UnsuccessfulException loginResp
+        -- If the response is non-null, the path is definitely a directory;
+        -- If the response is null, the path may be a file or may not exist.
+        dirList <- nlst handle [ "-a", path ]
+        when (BS.null dirList) $ do
+          fileSize <- size handle path `catch` \err ->
+            case err of
+              UnsuccessfulException _ -> throwError $ FtpEntryDoesNotExist path
+              FailureException FTPResponse{..} | frCode == 550 ->
+                throwError $ FtpEntryDoesNotExist path
+              _ -> liftIO $ throwIO err
+          when (fileSize == 0) $
+            throwError $ FtpEntryDoesNotExist path
```

Comment:
Could you leave a brief explanation as a comment here? Can't files in FTP really be empty and return `0` size?

Reply:
@balsoft do you have an answer?

Reply:
This is a mistake on my part. At least per the standard:

>    The server-PI will respond to the SIZE command with a 213 reply
>   giving the transfer size of the file whose pathname was supplied, or
>   an error response if the file does not exist, the size is
>   unavailable, or some other error has occurred.

So if we get the size without any exceptions, it should mean that the file exists (even if the size is 0).